### PR TITLE
replace `fmt::Write` with `io::Write`

### PR DESCRIPTION
**Pull request by daaku**
There are a couple of other changes as well, so I recommend a ff or merge commit, rather than squashing the changes together.

### CODE REVIEW

**Comment by Schniz** [comment]
> I'm not sure about the benefits of doing so

I'm not saying that it's not good, just that I don't really understand hehe 😄

**Comment by Schniz** [comment]
But from the looks of it, `io::Write` looks like a better fit

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -1,30 +1,17 @@
...
-      with:
-        rust-version: nightly
         components: rustfmt
     - uses: actions/checkout@v1
-    - name: cargo fmt
+    - name: Build
+      run: cargo build
+    - name: Test
+      run: cargo test
+    - name: Format
```
[the remark this answers is not in the document]
Comment (daaku) [reply .github/workflows/rust.yml:16]: > 
> 
> I think that merging them both will be actually slower: `cargo fmt` doesn't require the project to be built or the dependencies to be downloaded. Having them split would give a faster feedback loop

Pretty sure having 1 build job will be faster than having 2. Just generally good practice imho to have build, test, format, lint all in one job.

File: render_macros/src/lib.rs [hunk 2]
```diff
@@ -113,7 +113,7 @@ use syn::parse_macro_input;
 #[proc_macro_error]
 pub fn html(input: TokenStream) -> TokenStream {
     let el = proc_macro2::TokenStream::from(rsx(input));
-    let result = quote! { ::render::Render::render(#el) };
+    let result = quote! { ::render::render_to_string(#el) };
```
[the remark this answers is not in the document]
Comment (daaku) [reply render_macros/src/lib.rs:116]: Yeah, I wasn't sure if this macro was necessary. We can do what you suggested or drop the macro all together. The methods might be enough.

**Comment by daaku** [comment]
> But from the looks of it, `io::Write` looks like a better fit

Hehe, yes per rust docs seems like `fmt::Write` is kinda for internal-to-rust use. 😉

**Comment by daaku** [comment]
This request has ended up being everything since I based it off master and I've been hacking away at it as I see fit.

The most recent commit with support for raw text as children I think is possibly controversial because it has some specific rules around whitespace. The `proc_macro` tokenizer doesn't preserve whitespace, so the implementation makes some choices. I'm still playing around with it, but I think it should not throw users off, and has the nice effect of eliminating the need for writing things inside long braced + quoted strings.

**Comment by vpzomtrrfrt** [comment]
This seems like it should be split into multiple PRs

File: .github/workflows/rust.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (vpzomtrrfrt) [reply .github/workflows/rust.yml:16]: I assume having multiple jobs would make them run in parallel, I don't know how that would be slower

[the remark this answers is not in the document]
Comment (daaku) [reply .github/workflows/rust.yml:16]:
```sh
> time cargo fmt
cargo fmt  0.05s user 0.03s system 99% cpu 0.086 total
```

For a command that takes 86ms to run, spinning up another container job becomes mostly overhead which is difficult to eliminate. As for running them in parallel, sure, that may theoretically be true, but practically GitHub Actions involve queues, not to mention other overhead besides the things we declare to run inside a container.

**Comment by Schniz** [comment]
> The most recent commit with support for raw text as children I think is possibly controversial because it has some specific rules around whitespace

Let's not do this in this PR. I'm also not so sure that we would like that — it's nice but it can be much harder to parse.

Also, since the CI doesn't really work for forks (ah, so annoying) — let's revert these changes as well. Other than that, I really like the PR — and although it may be a breaking change, I really didn't know that std::fmt::Write is considered somewhat private. My bad!

**Comment by daaku** [comment]
> I'm also not so sure that we would like that — it's nice but it can be much harder to parse.

That's unfortunate.. I really find having to brace+quote all my text is a frustrating experience as a consumer of this library (same for attribute values). It's the first thing I ran into as I was trying to port one of my applications to use this library.

I'll try to split the `io::Write` bit into another pull request for you.

**Comment by Schniz** [comment]
> > I'm also not so sure that we would like that — it's nice but it can be much harder to parse.
> 
> 
> 
> That's unfortunate.. I really find having to brace+quote all my text is a frustrating experience as a consumer of this library (same for attribute values). It's the first thing I ran into as I was trying to port one of my applications to use this library.


Just to make sure - I’m not cancelling your opinion. I’m grateful for your contributions. Let’s make a discussion 😇 I will comment some insights when I won’t be AFK

**Comment by daaku** [comment]
Glad to know you're not outright opposed to it :smile:

**Comment by Schniz** [comment]
Haha no I don't 😃 but it's a big change so maybe others could chime in and say what they think as well.

My 2 cents:

In the original JSX implementation you could do that — and I guess this is why both Babel and TypeScript allow it. This is very convenient when you do a simple app or designing something or even copying HTML into your code. But:

* Whitespace is still an issue, although in JSX it goes like "parse all the words that have spaces between them, line breaks don't count" — so maybe that'll work 🤷‍♂️ 
* The Reason JSX syntax does not allow plain strings as children, and for many reasons it feels like JSX 2.0 — along with punning, that makes more sense (`<X pun />` => `<X pun={pun} />` and not `<X pun={true} />`). IIRC, The Reason community explained why it doesn't really change production apps by saying that "real world" apps tend to extract their language files, especially if you're in a RTL world (unfortunately, my first language _is_ RTL and most dev tools are only LTR friendly)
* The library right now tries to be as simple and straightforward that it can. The more features we bake into the macro, the more error prone it can get or get supported in other tools. This is why curly braces are required in arguments. We can relax that too — `<Component prop=true>` wouldn't be hard to parse as well. but it kinda easier on the eyes (and probably easier to write a syntax highlighter) when everything has the same format.

But again, the biggest pro here is that it will be much easier to use. I'm not sure about it. I'll tell you what — even if we will think that it is too hacky and we're not sure about it — we can add it under a feature flag. Then you can see how it works on your projects and eventually we can merge it upstream. 😄 

but let's do it one PR at a time 😈 

what do you think?

**Comment by daaku** [comment]
> In the original JSX implementation you could do that — and I guess this is why both Babel and TypeScript allow it. This is very convenient when you do a simple app or designing something or even copying HTML into your code. But:
> 
> Whitespace is still an issue, although in JSX it goes like "parse all the words that have spaces between them, line breaks don't count" — so maybe that'll work man_shrugging
> 

I don't know about the original implementation, but the currently JSX has it's own set of rules around whitespace (like trimming around newlines). They're just that I guess, a set of rules they decided on.

> The Reason JSX syntax does not allow plain strings as children, and for many reasons it feels like JSX 2.0 — along with punning, that makes more sense (`<X pun />` => `<X pun={pun} />` and not `<X pun={true} />`). IIRC, The Reason community explained why it doesn't really change production apps by saying that "real world" apps tend to extract their language files, especially if you're in a RTL world (unfortunately, my first language _is_ RTL and most dev tools are only LTR friendly)

I have no experience with Reason, but I think the argument for "real world" is not one that I like. Planning for language files and adding that complexity at the get-go feels like a hindrance to me. Making RTL work correctly I'm totally on board with, though I have few thoughts and little experience on how that typically works in an environment like JSX or when embedding markup in code. The ability of using language files and externalizing text doesn't go away when adding support for this obviously.

> The library right now tries to be as simple and straightforward that it can. The more features we bake into the macro, the more error prone it can get or get supported in other tools. This is why curly braces are required in arguments. We can relax that too — `<Component prop=true>` wouldn't be hard to parse as well. but it kinda easier on the eyes (and probably easier to write a syntax highlighter) when everything has the same format.

This is a balance for sure, and I would find it quite a pain if my tooling crapped out due to the formatting. I tend to use vscode and vim along with rust-analyzer, and so far they all are handling these changes (or at least nothing has crapped out on me yet!). You're also right there's a slippery slope of supporting things like `<Component prop=true>` -- I'm not sure where to draw the line to be honest.

> But again, the biggest pro here is that it will be much easier to use. I'm not sure about it. I'll tell you what — even if we will think that it is too hacky and we're not sure about it — we can add it under a feature flag. Then you can see how it works on your projects and eventually we can merge it upstream. :smile:

I'm going to keep hacking away and see how well it works, the overhead of managing this with a feature flag isn't something I want to deal with to be honest. If it ends up in a place that makes sense to you, then we can figure out how to merge things back. I haven't actually used this for anything I've deployed, but I'm a big fan of this JSX in Rust style approach and your work was a super solid starting point and I'm excited to see it be something I can build upon for my use case.

> but let's do it one PR at a time :smiling_imp:

That I for sure agree with!
